torch/csrc/jit/script/compiler.cpp

```diff
@@ -1333,6 +1331,54 @@ struct to_ir {
     }
   }
 
+  void emitIf(const If& stmt) {
+    // NOTE: emitIf checks on If stmt condition to see if the cond AST kind == is/is not,
+    // for such cases we do meta programming and disable emitting the corresponding branches
+    Expr cond = stmt.cond();
+
+    if (cond.kind() != TK_IS && cond.kind() != TK_ISNOT) {
+      // emit normal IF stmt for cases except TK_IS and TK_ISNOT
+      Value* cond_value = emitCond(cond);
+      emitIfElseBlocks(cond_value, stmt);
+    } else {
+      // meta programming on AST and emit branches base on the possible output of cond
+      auto cond_op = BinOp(cond);
+      SugaredValuePtr lhs_val = emitSugaredExpr(cond_op.lhs(), 1);
+      SugaredValuePtr rhs_val = emitSugaredExpr(cond_op.rhs(), 1);
+
+      List<Stmt> always_none_branch = cond.kind() == TK_IS? stmt.trueBranch(): stmt.falseBranch();
+      List<Stmt> never_none_branch = cond.kind() == TK_IS? stmt.falseBranch(): stmt.trueBranch();
+
+      if (lhs_val->isNone() && rhs_val->isNone()) {
+        // None is/is not None: only emit the always_none_branch
+        emitStatements(always_none_branch);
+      } else if (Value* v = asSimple(lhs_val)) {
+        if (v->type()->cast<OptionalType>()) {
+          // lhs_val maybe None: lhs value with optional type output value,
+          // emit the whole If stmt as usual, finish emitCond first
+          std::vector<NamedValue> named_values;
+          auto lhs_range = cond_op.lhs().get()->range();
+          auto rhs_range = cond_op.rhs().get()->range();
+          named_values.emplace_back(lhs_range, lhs_val->asValue(lhs_range, method));
+          named_values.emplace_back(rhs_range, rhs_val->asValue(rhs_range, method));
+          auto kind = getNodeKind(cond.kind(), cond.get()->trees().size());
+          Value* cond_value = emitBuiltinCall(
+                     cond.get()->range(),
+                     *method.graph(),
+                     kind,
+                     c10::nullopt,
+                     named_values,
+                     {},
+                     /*required=*/true);
+          emitIfElseBlocks(cond_value, stmt);
+        } else {
+          // lhs_val can never be none: only emit never_none_branch
```

Comment:
But the right-hand side might be none and that case is not handled.

Reply:
```
A A  => always_true_branch
A M  => emit conditional
A N  => always_false_branch
M A => emit conditional
M M => emit conditional
M  N => emit conditional
N A  => always_true branch
N M => emit conditional
N N => always true branch
```

```
if (either is maybe) emit conditional
else if (class1 == class2) emit true conditional else emit false conditional
```

Reply:
Shouldn't 

```
A M => always_true_branch
N A => always_false branch
```
Then group them will be:

AM, AA, -> always true
AN , NA-> always false
MA, MM, MN, NM, NN -> conditional

Reply:
Hmmm it is really confused how `A M` should behave since we don't take None as subtype of Optional Types, probably emit conditional should be fine.

Reply:
A M is dynamic the M may be a none but it may not be.

Reply:
Agreed, so the final dispatch logic is like the below:

AA, -> always_true_branch
AN , NA-> always_false branch
MA, MM, MN, NM, NN, AM -> emit both conditional branches